[glib] Document GRefPtr

ntrrgc opened the pull request:
#### 0a6b49d9681d1228b67093cd3f9e5d84642d06bf
<pre>
[glib] Document GRefPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=291896">https://bugs.webkit.org/show_bug.cgi?id=291896</a>

Reviewed by Michael Catanzaro and Carlos Garcia Campos.

Adds documentation to GRefPtr explaining how to use it and potential
pitfalls, including floating references.

* Source/WTF/wtf/glib/GRefPtr.h:
(WTF::GRefPtr::GRefPtr):
(WTF::GRefPtr::get const):
(WTF::adoptGRef):

Canonical link: <a href="https://commits.webkit.org/294108@main">https://commits.webkit.org/294108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/143b649362d13c6c9f6c62ed94f77e445ce39d6f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105866 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51317 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102770 "Build is in progress. Recent messages:") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28855 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76704 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33740 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103736 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15894 "Found 1 new test failure: fast/forms/ios/file-upload-panel-dismiss-when-view-removed-from-window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90985 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57057 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/100203 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15703 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8993 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50693 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93393 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108221 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99337 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27847 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20448 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85663 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28210 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87187 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85204 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21711 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29907 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7635 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21843 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27782 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33037 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122963 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27593 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34276 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30911 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29151 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos reviewed (approved):
Awesome!

TingPing commented on Source/WTF/wtf/glib/GRefPtr.h:
```diff
@@ -116,7 +140,10 @@ template <typename T> class GRefPtr {
+    // Borrows the raw pointer from GRefPtr.
+    // The pointer is guaranteed to be valid for as long as GRefPtr holds an owning reference
+    // to that object.
+    T* /* transfer none */ get() const { return m_ptr; }
```
```suggestion
    T* /* (transfer none) */ get() const { return m_ptr; }
```

Minor inconsistency.

calvaris commented on Source/WTF/wtf/glib/GRefPtr.h:
```diff
@@ -54,14 +54,37 @@ template <typename T> inline void derefGPtr(T*);
+    // This corresponds to the semantics of transfer floating in glib and g_object_ref_sink():
+    //
+    // * If ptr was a floating reference, it stops being floating and ownership *moves* into GRefPtr;
+    //   this corresponds to the same semantics as transfer full.
```
this corresponds to the same semantics as transfer floating, AFAIK

mcatanzaro replied:
The comment is correct: this is one bullet point explaining that transfer floating conditionally behaves as either transfer full or transfer none.

calvaris commented on Source/WTF/wtf/glib/GRefPtr.h:
```diff
@@ -116,7 +140,10 @@ template <typename T> class GRefPtr {
+    // to that object.
+    T* /* (transfer none) */ get() const { return m_ptr; }
     T& operator*() const { return *m_ptr; }
     ALWAYS_INLINE T* operator->() const { return m_ptr; }
```
I wonder if we could add LIFETIME_BOUND tags here even here or a new bug as this could be out of the scope of documenting

ntrrgc replied:
We could. It is technically possible to do .get() and safely keep the pointer after the GRefPtr is gone, but only if you have a guarantee that there will be at least one other strong reference holder while you use it... but relying on that is already a code smell.

Let's see what code gets flagged if I add it...

ntrrgc replied:
It builds, so I'm adding it in this patch.

calvaris commented on Source/WTF/wtf/glib/GRefPtr.h:
```diff
@@ -207,7 +234,15 @@ template <typename T> struct IsSmartPtr<GRefPtr<T>> {
+//
+// Specializations for types where floating references are common should contain debug assertions
+// guarding against accidental use with floating references.
+template <typename T> GRefPtr<T> adoptGRef(T* p /* (transfer full) */)
```
You could add here that this corresponds to transfer full.

ntrrgc replied:
It already says `/* (transfer full) */`, but I could mention it in the first paragraph too if that's what you meant.

calvaris reviewed (commented):
I would leave the final word on this to someone with more experience than me, like @carlosgcampos , but I made some comments.

ntrrgc committed:
[glib] Document GRefPtr
https://bugs.webkit.org/show_bug.cgi?id=291896

Reviewed by Michael Catanzaro and Carlos Garcia Campos.

Adds documentation to GRefPtr explaining how to use it and potential
pitfalls, including floating references.

* Source/WTF/wtf/glib/GRefPtr.h:
(WTF::GRefPtr::GRefPtr):
(WTF::GRefPtr::get const):
(WTF::adoptGRef):

Canonical link: https://commits.webkit.org/294108@main
